crates/shared/src/sources/balancer_v2/pool_cache.rs

```diff
@@ -88,27 +88,20 @@ impl CacheFetching<H256, WeightedPool> for PoolReserveFetcher {
             .await
             .into_iter()
             .map(|registered_pool| {
-                let (common_pool_state, common_pool_state_ok) =
-                    share_common_pool_state(self.common_pool_fetcher.fetch_common_pool_state(
-                        &registered_pool.common,
-                        &mut batch,
-                        block,
-                    ));
-                let weighted_pool_state = self.weighted_pool_factory.fetch_pool_state(
-                    &registered_pool,
-                    common_pool_state_ok.boxed(),
-                    &mut batch,
-                    block,
-                );
+                let pool_status = self
+                    .pool_fetcher
+                    .fetch_pool(&registered_pool, &mut batch, block);
 
                 async move {
-                    let common_pool_state = common_pool_state.await?;
-                    let weighted_pool_state = weighted_pool_state.await?;
-                    Ok(WeightedPool::new(
-                        registered_pool,
-                        common_pool_state,
-                        weighted_pool_state,
-                    ))
+                    let pool_status = pool_status.await?;
+                    let weighted_pool = match pool_status.active() {
+                        Some(Pool {
+                            kind: PoolKind::Weighted(state),
+                            ..
+                        }) => WeightedPool::new_unpaused(registered_pool, state),
```

Comment:
nanonit: I find it slightly easier to read when both return points are on the same branch level
```suggestion
                        }) => return Ok(Some(WeightedPool::new_unpaused(registered_pool, state))),
```